### PR TITLE
Support typings in publishConfig

### PR DESCRIPTION
**Pull request by bgotink**
Implements #137
Creates a basis for #133 to start from

To do:

- [x] write tests
- [x] `git rebase -i --autosquash`
- [x] &hellip;

### CODE REVIEW

File: packages/berry-fslib/package.json [hunk 1]
```diff
@@ -15,7 +15,8 @@
     "postpack": "rm -rf lib"
   },
   "publishConfig": {
-    "main": "./lib/index.js"
+    "main": "./lib/index.js",
+    "typings": "./lib/index.d.ts"
```
Comment (bgotink) [inline packages/berry-fslib/package.json]: It made sense for me to add typings to this package, feel free to disagree :) 

I wanted to add typings to `@berry/libzip` as well but it doesn't work out of the box due to the `.d.ts` file in the `sources` folder.

File: packages/plugin-pack/sources/index.ts [hunk 2]
```diff
@@ -10,9 +11,17 @@ export interface Hooks {
     project: Project,
     definePath: (path: string | null) => void,
   ) => Promise<void>,
+
+  beforeWorkspacePacking?: (
```
Comment (bgotink) [inline packages/plugin-pack/sources/index.ts]: In https://github.com/yarnpkg/berry/issues/137#issuecomment-490647425 you mentioned putting this hook or at least its implementation in `plugin-essentials`.
Imo it makes more sense to put the hook definition here, as the `packUtils` actually call the hook. I placed the main/module hook here as well, do you want me to move it to plugin-essentials instead?

File: packages/plugin-typescript/package.json [hunk 3]
```diff
@@ -5,7 +5,8 @@
   "dependencies": {
     "@berry/builder": "workspace:*",
     "@berry/core": "workspace:*",
-    "@berry/plugin-essentials": "workspace:*"
+    "@berry/plugin-essentials": "workspace:*",
+    "@berry/plugin-pack": "workspace:*"
```
Comment (bgotink) [inline packages/plugin-typescript/package.json]: I'm not sure I like having to depend on plugin-pack just for a typescript interface definition

Comment (arcanis) [reply packages/plugin-typescript/package.json]: It's a bit different from standard bundling strategies - the `@berry/builder` toolchain doesn't actually aggrgate the plugins together when building one of them; they're instead expected to be downloaded independently by the user - or in the case of `plugin-pack`, directly linked against the version of `plugin-pack` bundled within the standard bundle (it's a bit like if they were peer dependencies, in a sense).

File: packages/berry-fslib/package.json (hunk 1 above)
Comment (bgotink) [reply packages/berry-fslib/package.json]: Yes but no&hellip; The documentation (quoted below) says they do that for top-level `index.js` and top-level `index.d.ts` but they also say please mark it in the package manifest.

Also note they support `typings` and `types`, so best I'd best add the `types` to the hook as well

---

If your package has a main .js file, you will need to indicate the main declaration file in your package.json file as well. Set the types property to point to your bundled declaration file. For example:

{
    "name": "awesome",
    "author": "Vandelay Industries",
    "version": "1.0.0",
    "main": "./lib/main.js",
    "types": "./lib/main.d.ts"
}
Note that the "typings" field is synonymous with "types", and could be used as well.

Also note that if your main declaration file is named index.d.ts and lives at the root of the package (next to index.js) you do not need to mark the "types" property, though it is advisable to do so.

~ https://www.typescriptlang.org/docs/handbook/declaration-files/publishing.html

**Comment by arcanis** [comment]
Seems like there's a problem with Windows getting stuck on `pack` 🤔

**Pull request merged**